exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql

```diff
@@ -526,47 +527,135 @@ declare %private function test:map-named-argument($testArgs as element(value)*,
+        case "xs:string"
+        return
+            xs:string($annotation-value/text())
+
+        case "xs:integer"
+        return
+            xs:integer($annotation-value/text())
+
+        case "xs:decimal"
+        return
+            xs:decimal($annotation-value/text())
+
+        case "xs:double"
+        return
+            xs:double($annotation-value/text())
+
+        default
+        return
+            fn:error(
+                $test:UNKNOWN_ANNOTATION_VALUE_TYPE,
+                "The annotation value has an unknown type"
+            )
+};
+
+declare %private function test:cast($targs as element(value)*, $farg as element(argument)) {
     for $targ in $targs
+    let $varg := test:xdm-value-from-annotation-value($targ)
     return
         switch (string($farg/@type))
-            case "xs:string" return
-                string($targ)
-            case "xs:integer" case "xs:int" return
-                xs:integer($targ)
-            case "xs:decimal" return
-                xs:decimal($targ)
-            case "xs:float" case "xs:double" return
-                xs:double($targ)
-            case "xs:date" return
-                xs:date($targ)
-            case "xs:dateTime" return
-                xs:dateTime($targ)
-            case "xs:time" return
-                xs:time($targ)
-            case "element()" return
-                parse-xml($targ)/*
-            case "text()" return
-                text { string($targ) }
-            default return
-                $targ
+            case "xs:string"
+            return
+                if ($varg instance of xs:string)
+                then
+                    $varg
+                else
+                    string($varg)
+
+            case "xs:integer"
+            return
+                if ($varg instance of xs:integer)
+                then
+                    $varg
+                else
+                    xs:integer($varg)
+
+            case "xs:int"
+            return
+                xs:int($varg)
+
+            case "xs:decimal"
+            return
+                if ($varg instance of xs:decimal)
+                then
+                    $varg
+                else
+                    xs:decimal($varg)
+
+            case "xs:float"
+            return
+                xs:float($varg)
+
+            case "xs:double"
+            return
+                if ($varg instance of xs:decimal)
+                then
+                    $varg
+                else
+                    xs:double($varg)
+
+            case "xs:date"
+            return
+                xs:date($varg)
+
+            case "xs:dateTimeStamp"
+            return
+                xs:dateTimeStamp($varg)
+
+            case "xs:dateTime"
+            return
+                xs:dateTime($varg)
+
+            case "xs:time"
+            return
+                xs:time($varg)
+
+            case "element()"
+            return
+                fn:parse-xml($varg)/element()
```

Comment:
The `fn:parse-xml` function returns a `document-node`, so selecting `/element()` here will return an empty sequence. See https://www.w3.org/TR/xpath-functions-31/#func-parse-xml.

Reply:
I think this should work as I usually select the first child node() when dealing with document-nodes

Reply:
Ah, you're quite right. Please disregard my comment above.